{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 77 genomes in the dataset.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>n_lineages</th>\n",
       "      <th>lineage</th>\n",
       "      <th>sublineage</th>\n",
       "      <th>mycobacterial_reads</th>\n",
       "      <th>tb_reads</th>\n",
       "      <th>tb_coverage</th>\n",
       "      <th>tb_depth</th>\n",
       "      <th>mykrobe_lineage_coverage</th>\n",
       "      <th>mykrobe_lineage_depth</th>\n",
       "      <th>antibiogram</th>\n",
       "      <th>pipeline_build</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uniqueid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMN03648077.lab.1.iso.1</th>\n",
       "      <td>M. tuberculosis</td>\n",
       "      <td>1</td>\n",
       "      <td>lineage4</td>\n",
       "      <td>lineage4.6.1.2</td>\n",
       "      <td>4820884</td>\n",
       "      <td>4695252</td>\n",
       "      <td>99.7035</td>\n",
       "      <td>103.2910</td>\n",
       "      <td>100.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>RRSSSSUSSSRSUS</td>\n",
       "      <td>149a518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMEA1403592.lab.1.iso.1</th>\n",
       "      <td>M. tuberculosis</td>\n",
       "      <td>1</td>\n",
       "      <td>lineage4</td>\n",
       "      <td>lineage4.2.1</td>\n",
       "      <td>3290666</td>\n",
       "      <td>3213994</td>\n",
       "      <td>99.8773</td>\n",
       "      <td>69.0550</td>\n",
       "      <td>100.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>SSSSSSSSSSSSSS</td>\n",
       "      <td>149a518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMN07766210.lab.1.iso.1</th>\n",
       "      <td>M. tuberculosis</td>\n",
       "      <td>1</td>\n",
       "      <td>lineage1</td>\n",
       "      <td>lineage1.2.1</td>\n",
       "      <td>3667176</td>\n",
       "      <td>3585232</td>\n",
       "      <td>99.7362</td>\n",
       "      <td>99.2324</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SSSSSSUSSSRSUS</td>\n",
       "      <td>149a518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                species  n_lineages   lineage  \\\n",
       "uniqueid                                                                        \n",
       "site.ENA.subj.SAMN03648077.lab.1.iso.1  M. tuberculosis           1  lineage4   \n",
       "site.ENA.subj.SAMEA1403592.lab.1.iso.1  M. tuberculosis           1  lineage4   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1  M. tuberculosis           1  lineage1   \n",
       "\n",
       "                                            sublineage  mycobacterial_reads  \\\n",
       "uniqueid                                                                      \n",
       "site.ENA.subj.SAMN03648077.lab.1.iso.1  lineage4.6.1.2              4820884   \n",
       "site.ENA.subj.SAMEA1403592.lab.1.iso.1    lineage4.2.1              3290666   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1    lineage1.2.1              3667176   \n",
       "\n",
       "                                        tb_reads  tb_coverage  tb_depth  \\\n",
       "uniqueid                                                                  \n",
       "site.ENA.subj.SAMN03648077.lab.1.iso.1   4695252      99.7035  103.2910   \n",
       "site.ENA.subj.SAMEA1403592.lab.1.iso.1   3213994      99.8773   69.0550   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1   3585232      99.7362   99.2324   \n",
       "\n",
       "                                        mykrobe_lineage_coverage  \\\n",
       "uniqueid                                                           \n",
       "site.ENA.subj.SAMN03648077.lab.1.iso.1                     100.0   \n",
       "site.ENA.subj.SAMEA1403592.lab.1.iso.1                     100.0   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1                       0.0   \n",
       "\n",
       "                                        mykrobe_lineage_depth     antibiogram  \\\n",
       "uniqueid                                                                        \n",
       "site.ENA.subj.SAMN03648077.lab.1.iso.1                   90.0  RRSSSSUSSSRSUS   \n",
       "site.ENA.subj.SAMEA1403592.lab.1.iso.1                   64.0  SSSSSSSSSSSSSS   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1                    0.0  SSSSSSUSSSRSUS   \n",
       "\n",
       "                                       pipeline_build  \n",
       "uniqueid                                               \n",
       "site.ENA.subj.SAMN03648077.lab.1.iso.1        149a518  \n",
       "site.ENA.subj.SAMEA1403592.lab.1.iso.1        149a518  \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1        149a518  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes = pandas.read_parquet('genomes.parquet')\n",
    "print(f\"There are {len(genomes)} genomes in the dataset.\")\n",
    "genomes[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 77 entries, site.ENA.subj.SAMN03648077.lab.1.iso.1 to site.ENA.subj.SAMEA1016056.lab.1.iso.1\n",
      "Data columns (total 12 columns):\n",
      " #   Column                    Non-Null Count  Dtype   \n",
      "---  ------                    --------------  -----   \n",
      " 0   species                   77 non-null     object  \n",
      " 1   n_lineages                77 non-null     int64   \n",
      " 2   lineage                   77 non-null     category\n",
      " 3   sublineage                77 non-null     category\n",
      " 4   mycobacterial_reads       77 non-null     int64   \n",
      " 5   tb_reads                  77 non-null     int64   \n",
      " 6   tb_coverage               77 non-null     float64 \n",
      " 7   tb_depth                  77 non-null     float64 \n",
      " 8   mykrobe_lineage_coverage  76 non-null     float64 \n",
      " 9   mykrobe_lineage_depth     76 non-null     float64 \n",
      " 10  antibiogram               77 non-null     category\n",
      " 11  pipeline_build            77 non-null     category\n",
      "dtypes: category(4), float64(4), int64(3), object(1)\n",
      "memory usage: 8.7+ KB\n"
     ]
    }
   ],
   "source": [
    "genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>center_name</th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>first_public</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>fastq_md5</th>\n",
       "      <th>fastq_bytes</th>\n",
       "      <th>has_effects</th>\n",
       "      <th>has_variants</th>\n",
       "      <th>has_mutations</th>\n",
       "      <th>has_predictions</th>\n",
       "      <th>has_main_report</th>\n",
       "      <th>in_mapping_file</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PTAN-0001.lab.TAN-001.iso.1</th>\n",
       "      <td>PRJEB23245</td>\n",
       "      <td>SAMEA104362043</td>\n",
       "      <td>ERR2184202</td>\n",
       "      <td>University of Oxford;Nuffield Department of Me...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2019-06-12</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/002/ERR218...</td>\n",
       "      <td>612e42126398723aa1ee84dd1e06feec;349b4db6faead...</td>\n",
       "      <td>79400918;105468452</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PTAN-0002.lab.TAN-004.iso.1</th>\n",
       "      <td>PRJEB23245</td>\n",
       "      <td>SAMEA104362044</td>\n",
       "      <td>ERR2184203</td>\n",
       "      <td>University of Oxford;Nuffield Department of Me...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2019-06-12</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/003/ERR218...</td>\n",
       "      <td>469becc8e8fb3961bde32b5352858190;b13aca3a9005c...</td>\n",
       "      <td>67987132;86536083</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.05.subj.PTAN-0003.lab.TAN-006.iso.1</th>\n",
       "      <td>PRJEB23245</td>\n",
       "      <td>SAMEA104362045</td>\n",
       "      <td>ERR2184204</td>\n",
       "      <td>University of Oxford;Nuffield Department of Me...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2019-06-12</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/004/ERR218...</td>\n",
       "      <td>d97d6b0ae3ba59f6bfa19e304ac8d94f;022a232fe2db3...</td>\n",
       "      <td>57530779;72231695</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         study_accession sample_accession  \\\n",
       "UNIQUEID                                                                    \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1      PRJEB23245   SAMEA104362043   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1      PRJEB23245   SAMEA104362044   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1      PRJEB23245   SAMEA104362045   \n",
       "\n",
       "                                         run_accession  \\\n",
       "UNIQUEID                                                 \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1    ERR2184202   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1    ERR2184203   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1    ERR2184204   \n",
       "\n",
       "                                                                                center_name  \\\n",
       "UNIQUEID                                                                                      \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1  University of Oxford;Nuffield Department of Me...   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1  University of Oxford;Nuffield Department of Me...   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1  University of Oxford;Nuffield Department of Me...   \n",
       "\n",
       "                                         country location first_public  \\\n",
       "UNIQUEID                                                                 \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1    None     None   2019-06-12   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1    None     None   2019-06-12   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1    None     None   2019-06-12   \n",
       "\n",
       "                                                                                  fastq_ftp  \\\n",
       "UNIQUEID                                                                                      \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/002/ERR218...   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/003/ERR218...   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR218/004/ERR218...   \n",
       "\n",
       "                                                                                  fastq_md5  \\\n",
       "UNIQUEID                                                                                      \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1  612e42126398723aa1ee84dd1e06feec;349b4db6faead...   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1  469becc8e8fb3961bde32b5352858190;b13aca3a9005c...   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1  d97d6b0ae3ba59f6bfa19e304ac8d94f;022a232fe2db3...   \n",
       "\n",
       "                                                 fastq_bytes  has_effects  \\\n",
       "UNIQUEID                                                                    \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1  79400918;105468452        False   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1   67987132;86536083        False   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1   57530779;72231695        False   \n",
       "\n",
       "                                          has_variants  has_mutations  \\\n",
       "UNIQUEID                                                                \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1         False          False   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1         False          False   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1         False          False   \n",
       "\n",
       "                                          has_predictions  has_main_report  \\\n",
       "UNIQUEID                                                                     \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1            False            False   \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1            False            False   \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1            False            False   \n",
       "\n",
       "                                          in_mapping_file  \n",
       "UNIQUEID                                                   \n",
       "site.05.subj.PTAN-0001.lab.TAN-001.iso.1            False  \n",
       "site.05.subj.PTAN-0002.lab.TAN-004.iso.1            False  \n",
       "site.05.subj.PTAN-0003.lab.TAN-006.iso.1            False  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master = pandas.read_parquet('ENA_LOOKUP.parquet')\n",
    "master[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>center_name</th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>first_public</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>fastq_md5</th>\n",
       "      <th>fastq_bytes</th>\n",
       "      <th>has_effects</th>\n",
       "      <th>has_variants</th>\n",
       "      <th>has_mutations</th>\n",
       "      <th>has_predictions</th>\n",
       "      <th>has_main_report</th>\n",
       "      <th>in_mapping_file</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site.04.subj.00042.lab.631456.iso.1</th>\n",
       "      <td>PRJEB41116</td>\n",
       "      <td>SAMEA7524735</td>\n",
       "      <td>ERR4796279</td>\n",
       "      <td>University of Oxford;Nuffield Department of Me...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2021-10-20</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR479/009/ERR479...</td>\n",
       "      <td>be8d145568d28e534de15e2157294d4a;5e8ca8f43328b...</td>\n",
       "      <td>156611375;143692824</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMEA104394706.lab.1.iso.1</th>\n",
       "      <td>PRJEB9680</td>\n",
       "      <td>SAMEA104394706</td>\n",
       "      <td>ERR2200134</td>\n",
       "      <td>FORSCHUNGSZENTRUM BORSTEL;FZB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2017-11-21</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR220/004/ERR220...</td>\n",
       "      <td>eb298f759ee4949a0c3f4c1de2869dd7;71f0518f14d40...</td>\n",
       "      <td>180077755;189046222</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMN08215132.lab.1.iso.1</th>\n",
       "      <td>PRJNA413593</td>\n",
       "      <td>SAMN08215132</td>\n",
       "      <td>SRR6397368</td>\n",
       "      <td>SUB3191708</td>\n",
       "      <td>Canada: British Columbia</td>\n",
       "      <td>49.15 N 123.7 W</td>\n",
       "      <td>2018-03-03</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR639/008/SRR639...</td>\n",
       "      <td>4afbb981adc7618fcdbc46230fcb048e;a134e77f93071...</td>\n",
       "      <td>267175447;276979986</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMN03647408.lab.1.iso.1</th>\n",
       "      <td>PRJNA282721</td>\n",
       "      <td>SAMN03647408</td>\n",
       "      <td>SRR2100235</td>\n",
       "      <td>None</td>\n",
       "      <td>missing</td>\n",
       "      <td>None</td>\n",
       "      <td>2015-07-26</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR210/005/SRR210...</td>\n",
       "      <td>6eb5428d419bb4259dd096a7feea8da6;2787c02998af8...</td>\n",
       "      <td>354956962;355236453</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMN03647375.lab.1.iso.1</th>\n",
       "      <td>PRJNA282721</td>\n",
       "      <td>SAMN03647375</td>\n",
       "      <td>SRR2100202</td>\n",
       "      <td>None</td>\n",
       "      <td>missing</td>\n",
       "      <td>None</td>\n",
       "      <td>2015-11-03</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR210/002/SRR210...</td>\n",
       "      <td>3a323a76414b48ab426ede4a33d37b80;3dda1fbe5b342...</td>\n",
       "      <td>224809798;229867065</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMEA2534732.lab.1.iso.1</th>\n",
       "      <td>PRJEB9680</td>\n",
       "      <td>SAMEA2534732</td>\n",
       "      <td>ERR552560</td>\n",
       "      <td>FORSCHUNGSZENTRUM BORSTEL;FZB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2015-06-24</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR552/ERR552560/...</td>\n",
       "      <td>f5a88e4874d01658869ac9e73e876f17;985f6adb1819a...</td>\n",
       "      <td>113661953;142846965</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMEA2534732.lab.1.iso.1</th>\n",
       "      <td>PRJEB9680</td>\n",
       "      <td>SAMEA2534732</td>\n",
       "      <td>ERR552559</td>\n",
       "      <td>FORSCHUNGSZENTRUM BORSTEL;FZB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2015-06-24</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR552/ERR552559/...</td>\n",
       "      <td>75ea182a20281bda59d2f590108d685d;56c2740995b72...</td>\n",
       "      <td>34612663;34857010</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMEA2534732.lab.1.iso.1</th>\n",
       "      <td>PRJEB9680</td>\n",
       "      <td>SAMEA2534732</td>\n",
       "      <td>ERR552561</td>\n",
       "      <td>FORSCHUNGSZENTRUM BORSTEL;FZB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2015-06-24</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR552/ERR552561/...</td>\n",
       "      <td>ca84fba5f5df4ed85e3f8d48a2b0f740;c2fbc37fa65ac...</td>\n",
       "      <td>35322473;35483515</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMN07766100.lab.1.iso.1</th>\n",
       "      <td>PRJNA413593</td>\n",
       "      <td>SAMN07766100</td>\n",
       "      <td>SRR6152831</td>\n",
       "      <td>None</td>\n",
       "      <td>Canada: British Columbia</td>\n",
       "      <td>49.15 N 123.7 W</td>\n",
       "      <td>2017-10-11</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR615/001/SRR615...</td>\n",
       "      <td>23f6286ff2f3ae660697e5468eeb288a;571d20fa0d560...</td>\n",
       "      <td>196482375;200658770</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>site.ENA.subj.SAMN07766210.lab.1.iso.1</th>\n",
       "      <td>PRJNA413593</td>\n",
       "      <td>SAMN07766210</td>\n",
       "      <td>SRR6153023</td>\n",
       "      <td>None</td>\n",
       "      <td>Canada: British Columbia</td>\n",
       "      <td>49.15 N 123.7 W</td>\n",
       "      <td>2017-10-11</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR615/003/SRR615...</td>\n",
       "      <td>50947e7f61814f3a1aa46d07f65efad5;02504e15f4ae0...</td>\n",
       "      <td>185831241;192517261</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         study_accession sample_accession  \\\n",
       "UNIQUEID                                                                    \n",
       "site.04.subj.00042.lab.631456.iso.1           PRJEB41116     SAMEA7524735   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1       PRJEB9680   SAMEA104394706   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1       PRJNA413593     SAMN08215132   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1       PRJNA282721     SAMN03647408   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1       PRJNA282721     SAMN03647375   \n",
       "...                                                  ...              ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1         PRJEB9680     SAMEA2534732   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1         PRJEB9680     SAMEA2534732   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1         PRJEB9680     SAMEA2534732   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1       PRJNA413593     SAMN07766100   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1       PRJNA413593     SAMN07766210   \n",
       "\n",
       "                                         run_accession  \\\n",
       "UNIQUEID                                                 \n",
       "site.04.subj.00042.lab.631456.iso.1         ERR4796279   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1    ERR2200134   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1      SRR6397368   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1      SRR2100235   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1      SRR2100202   \n",
       "...                                                ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1       ERR552560   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1       ERR552559   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1       ERR552561   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1      SRR6152831   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1      SRR6153023   \n",
       "\n",
       "                                                                                center_name  \\\n",
       "UNIQUEID                                                                                      \n",
       "site.04.subj.00042.lab.631456.iso.1       University of Oxford;Nuffield Department of Me...   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1                      FORSCHUNGSZENTRUM BORSTEL;FZB   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1                                           SUB3191708   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1                                                 None   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1                                                 None   \n",
       "...                                                                                     ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1                        FORSCHUNGSZENTRUM BORSTEL;FZB   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1                        FORSCHUNGSZENTRUM BORSTEL;FZB   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1                        FORSCHUNGSZENTRUM BORSTEL;FZB   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1                                                 None   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1                                                 None   \n",
       "\n",
       "                                                           country  \\\n",
       "UNIQUEID                                                             \n",
       "site.04.subj.00042.lab.631456.iso.1                           None   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1                      None   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1    Canada: British Columbia   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1                     missing   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1                     missing   \n",
       "...                                                            ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1                        None   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1                        None   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1                        None   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1    Canada: British Columbia   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1    Canada: British Columbia   \n",
       "\n",
       "                                                 location first_public  \\\n",
       "UNIQUEID                                                                 \n",
       "site.04.subj.00042.lab.631456.iso.1                  None   2021-10-20   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1             None   2017-11-21   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1    49.15 N 123.7 W   2018-03-03   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1               None   2015-07-26   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1               None   2015-11-03   \n",
       "...                                                   ...          ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1               None   2015-06-24   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1               None   2015-06-24   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1               None   2015-06-24   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1    49.15 N 123.7 W   2017-10-11   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1    49.15 N 123.7 W   2017-10-11   \n",
       "\n",
       "                                                                                  fastq_ftp  \\\n",
       "UNIQUEID                                                                                      \n",
       "site.04.subj.00042.lab.631456.iso.1       ftp.sra.ebi.ac.uk/vol1/fastq/ERR479/009/ERR479...   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR220/004/ERR220...   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1    ftp.sra.ebi.ac.uk/vol1/fastq/SRR639/008/SRR639...   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1    ftp.sra.ebi.ac.uk/vol1/fastq/SRR210/005/SRR210...   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1    ftp.sra.ebi.ac.uk/vol1/fastq/SRR210/002/SRR210...   \n",
       "...                                                                                     ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1    ftp.sra.ebi.ac.uk/vol1/fastq/ERR552/ERR552560/...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1    ftp.sra.ebi.ac.uk/vol1/fastq/ERR552/ERR552559/...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1    ftp.sra.ebi.ac.uk/vol1/fastq/ERR552/ERR552561/...   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1    ftp.sra.ebi.ac.uk/vol1/fastq/SRR615/001/SRR615...   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1    ftp.sra.ebi.ac.uk/vol1/fastq/SRR615/003/SRR615...   \n",
       "\n",
       "                                                                                  fastq_md5  \\\n",
       "UNIQUEID                                                                                      \n",
       "site.04.subj.00042.lab.631456.iso.1       be8d145568d28e534de15e2157294d4a;5e8ca8f43328b...   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1  eb298f759ee4949a0c3f4c1de2869dd7;71f0518f14d40...   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1    4afbb981adc7618fcdbc46230fcb048e;a134e77f93071...   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1    6eb5428d419bb4259dd096a7feea8da6;2787c02998af8...   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1    3a323a76414b48ab426ede4a33d37b80;3dda1fbe5b342...   \n",
       "...                                                                                     ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1    f5a88e4874d01658869ac9e73e876f17;985f6adb1819a...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1    75ea182a20281bda59d2f590108d685d;56c2740995b72...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1    ca84fba5f5df4ed85e3f8d48a2b0f740;c2fbc37fa65ac...   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1    23f6286ff2f3ae660697e5468eeb288a;571d20fa0d560...   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1    50947e7f61814f3a1aa46d07f65efad5;02504e15f4ae0...   \n",
       "\n",
       "                                                  fastq_bytes  has_effects  \\\n",
       "UNIQUEID                                                                     \n",
       "site.04.subj.00042.lab.631456.iso.1       156611375;143692824         True   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1  180077755;189046222         True   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1    267175447;276979986         True   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1    354956962;355236453         True   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1    224809798;229867065         True   \n",
       "...                                                       ...          ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1    113661953;142846965         True   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1      34612663;34857010         True   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1      35322473;35483515         True   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1    196482375;200658770         True   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1    185831241;192517261         True   \n",
       "\n",
       "                                          has_variants  has_mutations  \\\n",
       "UNIQUEID                                                                \n",
       "site.04.subj.00042.lab.631456.iso.1               True           True   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1          True           True   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1            True           True   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1            True           True   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1            True           True   \n",
       "...                                                ...            ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1            True           True   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1            True           True   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1            True           True   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1            True           True   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1            True           True   \n",
       "\n",
       "                                          has_predictions  has_main_report  \\\n",
       "UNIQUEID                                                                     \n",
       "site.04.subj.00042.lab.631456.iso.1                  True             True   \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1             True             True   \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1               True             True   \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1               True             True   \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1               True             True   \n",
       "...                                                   ...              ...   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1               True             True   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1               True             True   \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1               True             True   \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1               True             True   \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1               True             True   \n",
       "\n",
       "                                          in_mapping_file  \n",
       "UNIQUEID                                                   \n",
       "site.04.subj.00042.lab.631456.iso.1                 False  \n",
       "site.ENA.subj.SAMEA104394706.lab.1.iso.1            False  \n",
       "site.ENA.subj.SAMN08215132.lab.1.iso.1              False  \n",
       "site.ENA.subj.SAMN03647408.lab.1.iso.1              False  \n",
       "site.ENA.subj.SAMN03647375.lab.1.iso.1              False  \n",
       "...                                                   ...  \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1              False  \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1              False  \n",
       "site.ENA.subj.SAMEA2534732.lab.1.iso.1              False  \n",
       "site.ENA.subj.SAMN07766100.lab.1.iso.1              False  \n",
       "site.ENA.subj.SAMN07766210.lab.1.iso.1              False  \n",
       "\n",
       "[82 rows x 16 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master[master.has_main_report]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "45420"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master.index.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>gene</th>\n",
       "      <th>mutation</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>nucleotide_number</th>\n",
       "      <th>nucleotide_index</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>codes_protein</th>\n",
       "      <th>indel_length</th>\n",
       "      <th>indel_nucleotides</th>\n",
       "      <th>amino_acid_number</th>\n",
       "      <th>amino_acid_sequence</th>\n",
       "      <th>number_nucleotide_changes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE1</td>\n",
       "      <td>L485L</td>\n",
       "      <td>ctg</td>\n",
       "      <td>ttg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>485.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>485.0</td>\n",
       "      <td>L</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE14</td>\n",
       "      <td>A106A</td>\n",
       "      <td>gca</td>\n",
       "      <td>gcc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>106.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>106.0</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE16</td>\n",
       "      <td>A96A</td>\n",
       "      <td>gcg</td>\n",
       "      <td>gct</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>96.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>96.0</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 uniqueid  gene mutation  ref  alt  \\\n",
       "0  site.ENA.subj.SAMN07766182.lab.1.iso.1   PE1    L485L  ctg  ttg   \n",
       "1  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE14    A106A  gca  gcc   \n",
       "2  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE16     A96A  gcg  gct   \n",
       "\n",
       "   nucleotide_number  nucleotide_index  gene_position  codes_protein  \\\n",
       "0                NaN               NaN          485.0           True   \n",
       "1                NaN               NaN          106.0           True   \n",
       "2                NaN               NaN           96.0           True   \n",
       "\n",
       "   indel_length indel_nucleotides  amino_acid_number amino_acid_sequence  \\\n",
       "0           NaN              None              485.0                   L   \n",
       "1           NaN              None              106.0                   A   \n",
       "2           NaN              None               96.0                   A   \n",
       "\n",
       "   number_nucleotide_changes  \n",
       "0                        1.0  \n",
       "1                        1.0  \n",
       "2                        1.0  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations = pandas.read_parquet('mutations.parquet')\n",
    "# df.reset_index(inplace=True)\n",
    "mutations[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 116853 entries, 0 to 2993\n",
      "Data columns (total 14 columns):\n",
      " #   Column                     Non-Null Count   Dtype  \n",
      "---  ------                     --------------   -----  \n",
      " 0   uniqueid                   116853 non-null  object \n",
      " 1   gene                       116853 non-null  object \n",
      " 2   mutation                   116853 non-null  object \n",
      " 3   ref                        106043 non-null  object \n",
      " 4   alt                        106043 non-null  object \n",
      " 5   nucleotide_number          9017 non-null    float64\n",
      " 6   nucleotide_index           9017 non-null    float64\n",
      " 7   gene_position              116670 non-null  float64\n",
      " 8   codes_protein              116853 non-null  bool   \n",
      " 9   indel_length               10598 non-null   float64\n",
      " 10  indel_nucleotides          10598 non-null   object \n",
      " 11  amino_acid_number          99221 non-null   float64\n",
      " 12  amino_acid_sequence        99221 non-null   object \n",
      " 13  number_nucleotide_changes  106043 non-null  float64\n",
      "dtypes: bool(1), float64(6), object(7)\n",
      "memory usage: 12.6+ MB\n"
     ]
    }
   ],
   "source": [
    "mutations.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mutation\n",
       "a-3g        155\n",
       "A111A       146\n",
       "L63L        144\n",
       "A72A        125\n",
       "A18A        122\n",
       "           ... \n",
       "Q553E:65      1\n",
       "T285T:24      1\n",
       "I190T:55      1\n",
       "H255Y:30      1\n",
       "V145V         1\n",
       "Name: count, Length: 32541, dtype: int64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations.mutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>gene</th>\n",
       "      <th>mutation</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>nucleotide_number</th>\n",
       "      <th>nucleotide_index</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>codes_protein</th>\n",
       "      <th>indel_length</th>\n",
       "      <th>indel_nucleotides</th>\n",
       "      <th>amino_acid_number</th>\n",
       "      <th>amino_acid_sequence</th>\n",
       "      <th>number_nucleotide_changes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE35</td>\n",
       "      <td>E99!:15</td>\n",
       "      <td>gaa</td>\n",
       "      <td>taa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>99.0</td>\n",
       "      <td>!</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>A180A:10</td>\n",
       "      <td>gcc</td>\n",
       "      <td>gcg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>180.0</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>R225G:22</td>\n",
       "      <td>agg</td>\n",
       "      <td>ggg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>225.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>225.0</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>G509G:3</td>\n",
       "      <td>ggc</td>\n",
       "      <td>ggt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>509.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>509.0</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>A518G:4</td>\n",
       "      <td>gcc</td>\n",
       "      <td>ggc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>518.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>518.0</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2511</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>T794I:6</td>\n",
       "      <td>acc</td>\n",
       "      <td>atc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>794.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>794.0</td>\n",
       "      <td>I</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2684</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>pks12</td>\n",
       "      <td>A3586V:46</td>\n",
       "      <td>gcg</td>\n",
       "      <td>gtg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3586.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>3586.0</td>\n",
       "      <td>V</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2685</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>pks12</td>\n",
       "      <td>G3602G:60</td>\n",
       "      <td>ggt</td>\n",
       "      <td>ggc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3602.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>3602.0</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2720</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>plcC</td>\n",
       "      <td>Y97N:3</td>\n",
       "      <td>tac</td>\n",
       "      <td>aac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>97.0</td>\n",
       "      <td>N</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2854</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>rpfA</td>\n",
       "      <td>A209A:4</td>\n",
       "      <td>gca</td>\n",
       "      <td>gcg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>209.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>209.0</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17453 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      uniqueid       gene   mutation  ref  \\\n",
       "9       site.ENA.subj.SAMN07766182.lab.1.iso.1       PE35    E99!:15  gaa   \n",
       "21      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10   A180A:10  gcc   \n",
       "22      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10   R225G:22  agg   \n",
       "23      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    G509G:3  ggc   \n",
       "24      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A518G:4  gcc   \n",
       "...                                        ...        ...        ...  ...   \n",
       "2511  site.ENA.subj.SAMEA104394448.lab.1.iso.1      mmpL5    T794I:6  acc   \n",
       "2684  site.ENA.subj.SAMEA104394448.lab.1.iso.1      pks12  A3586V:46  gcg   \n",
       "2685  site.ENA.subj.SAMEA104394448.lab.1.iso.1      pks12  G3602G:60  ggt   \n",
       "2720  site.ENA.subj.SAMEA104394448.lab.1.iso.1       plcC     Y97N:3  tac   \n",
       "2854  site.ENA.subj.SAMEA104394448.lab.1.iso.1       rpfA    A209A:4  gca   \n",
       "\n",
       "      alt  nucleotide_number  nucleotide_index  gene_position  codes_protein  \\\n",
       "9     taa                NaN               NaN           99.0           True   \n",
       "21    gcg                NaN               NaN          180.0           True   \n",
       "22    ggg                NaN               NaN          225.0           True   \n",
       "23    ggt                NaN               NaN          509.0           True   \n",
       "24    ggc                NaN               NaN          518.0           True   \n",
       "...   ...                ...               ...            ...            ...   \n",
       "2511  atc                NaN               NaN          794.0           True   \n",
       "2684  gtg                NaN               NaN         3586.0           True   \n",
       "2685  ggc                NaN               NaN         3602.0           True   \n",
       "2720  aac                NaN               NaN           97.0           True   \n",
       "2854  gcg                NaN               NaN          209.0           True   \n",
       "\n",
       "      indel_length indel_nucleotides  amino_acid_number amino_acid_sequence  \\\n",
       "9              NaN              None               99.0                   !   \n",
       "21             NaN              None              180.0                   A   \n",
       "22             NaN              None              225.0                   G   \n",
       "23             NaN              None              509.0                   G   \n",
       "24             NaN              None              518.0                   G   \n",
       "...            ...               ...                ...                 ...   \n",
       "2511           NaN              None              794.0                   I   \n",
       "2684           NaN              None             3586.0                   V   \n",
       "2685           NaN              None             3602.0                   G   \n",
       "2720           NaN              None               97.0                   N   \n",
       "2854           NaN              None              209.0                   A   \n",
       "\n",
       "      number_nucleotide_changes  \n",
       "9                           1.0  \n",
       "21                          1.0  \n",
       "22                          1.0  \n",
       "23                          1.0  \n",
       "24                          1.0  \n",
       "...                         ...  \n",
       "2511                        1.0  \n",
       "2684                        1.0  \n",
       "2685                        1.0  \n",
       "2720                        1.0  \n",
       "2854                        1.0  \n",
       "\n",
       "[17453 rows x 14 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations[mutations.mutation.str.contains(':')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>codon_idx</th>\n",
       "      <th>nucleotide_index</th>\n",
       "      <th>indel_length</th>\n",
       "      <th>indel_nucleotides</th>\n",
       "      <th>vcf_evidence</th>\n",
       "      <th>vcf_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>1977a&gt;g</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>-75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1977</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"DP\": 115, \"GT\": [1, 1], \"GT_CONF\": 738.27, \"...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>4013t&gt;c</td>\n",
       "      <td>recF</td>\n",
       "      <td>245.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4013</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"ALLELE_DP\": [0, 110], \"COV\": [0, 110], \"DP\":...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>7362g&gt;c</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7362</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"ALLELE_DP\": [0, 127], \"GT\": [1, 1], \"DP\": 12...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 uniqueid  variant  gene  gene_position  \\\n",
       "0  site.ENA.subj.SAMN01908825.lab.1.iso.1  1977a>g  dnaN          -75.0   \n",
       "1  site.ENA.subj.SAMN01908825.lab.1.iso.1  4013t>c  recF          245.0   \n",
       "2  site.ENA.subj.SAMN01908825.lab.1.iso.1  7362g>c  gyrA           21.0   \n",
       "\n",
       "   codon_idx  nucleotide_index  indel_length indel_nucleotides  \\\n",
       "0        NaN              1977             0              None   \n",
       "1        1.0              4013             0              None   \n",
       "2        0.0              7362             0              None   \n",
       "\n",
       "                                        vcf_evidence  vcf_idx  \n",
       "0  {\"DP\": 115, \"GT\": [1, 1], \"GT_CONF\": 738.27, \"...      1.0  \n",
       "1  {\"ALLELE_DP\": [0, 110], \"COV\": [0, 110], \"DP\":...      1.0  \n",
       "2  {\"ALLELE_DP\": [0, 127], \"GT\": [1, 1], \"DP\": 12...      1.0  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants = pandas.read_parquet('variants.parquet')\n",
    "variants[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>NUCLEOTIDE_INDEX</th>\n",
       "      <th>GENE_POSITION</th>\n",
       "      <th>CODES_PROTEIN</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_NUCLEOTIDES</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>AMINO_ACID_SEQUENCE</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>MINOR_MUTATION</th>\n",
       "      <th>FRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.01.subj.DR0682.lab.DR0682.iso.1</td>\n",
       "      <td>fadD34</td>\n",
       "      <td>S16W</td>\n",
       "      <td>tcg</td>\n",
       "      <td>tgg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>16.0</td>\n",
       "      <td>W</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.01.subj.DR0682.lab.DR0682.iso.1</td>\n",
       "      <td>fadD34</td>\n",
       "      <td>T26S</td>\n",
       "      <td>act</td>\n",
       "      <td>tct</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>26.0</td>\n",
       "      <td>S</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.01.subj.DR0682.lab.DR0682.iso.1</td>\n",
       "      <td>Rv1231c</td>\n",
       "      <td>L60L</td>\n",
       "      <td>ctg</td>\n",
       "      <td>ctc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>60.0</td>\n",
       "      <td>L</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               UNIQUEID     GENE MUTATION  REF  ALT  \\\n",
       "0  site.01.subj.DR0682.lab.DR0682.iso.1   fadD34     S16W  tcg  tgg   \n",
       "1  site.01.subj.DR0682.lab.DR0682.iso.1   fadD34     T26S  act  tct   \n",
       "2  site.01.subj.DR0682.lab.DR0682.iso.1  Rv1231c     L60L  ctg  ctc   \n",
       "\n",
       "   NUCLEOTIDE_NUMBER  NUCLEOTIDE_INDEX  GENE_POSITION  CODES_PROTEIN  \\\n",
       "0                NaN               NaN           16.0           True   \n",
       "1                NaN               NaN           26.0           True   \n",
       "2                NaN               NaN           60.0           True   \n",
       "\n",
       "   INDEL_LENGTH INDEL_NUCLEOTIDES  AMINO_ACID_NUMBER AMINO_ACID_SEQUENCE  \\\n",
       "0           NaN               nan               16.0                   W   \n",
       "1           NaN               nan               26.0                   S   \n",
       "2           NaN               nan               60.0                   L   \n",
       "\n",
       "   NUMBER_NUCLEOTIDE_CHANGES  IS_MINOR_ALLELE MINOR_MUTATION  FRS  \n",
       "0                        1.0            False           None  NaN  \n",
       "1                        1.0            False           None  NaN  \n",
       "2                        1.0            False           None  NaN  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS = pandas.read_pickle('/Users/fowler/files/cryptic/cryptic-tables-v2.1.2/MUTATIONS.pkl.gz')\n",
    "MUTATIONS.reset_index(inplace=True)\n",
    "MUTATIONS[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>VARIANT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>GENE_POSITION</th>\n",
       "      <th>CODON_IDX</th>\n",
       "      <th>NUCLEOTIDE_INDEX</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_NUCLEOTIDES</th>\n",
       "      <th>VCF_EVIDENCE</th>\n",
       "      <th>VCF_IDX</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>MINOR_VARIANT</th>\n",
       "      <th>FRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.01.subj.DR0682.lab.DR0682.iso.1</td>\n",
       "      <td>1977a&gt;g</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>-75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1977.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [1, 1], \"DP\": 45.0, \"ALLELE_DP\": [0.0, ...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.01.subj.DR0682.lab.DR0682.iso.1</td>\n",
       "      <td>4013t&gt;c</td>\n",
       "      <td>recF</td>\n",
       "      <td>245.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4013.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [1, 1], \"DP\": 103.0, \"ALLELE_DP\": [0.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.01.subj.DR0682.lab.DR0682.iso.1</td>\n",
       "      <td>7362g&gt;c</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7362.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [1, 1], \"DP\": 61.0, \"ALLELE_DP\": [0.0, ...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               UNIQUEID  VARIANT  GENE  GENE_POSITION  \\\n",
       "0  site.01.subj.DR0682.lab.DR0682.iso.1  1977a>g  dnaN          -75.0   \n",
       "1  site.01.subj.DR0682.lab.DR0682.iso.1  4013t>c  recF          245.0   \n",
       "2  site.01.subj.DR0682.lab.DR0682.iso.1  7362g>c  gyrA           21.0   \n",
       "\n",
       "   CODON_IDX  NUCLEOTIDE_INDEX  INDEL_LENGTH INDEL_NUCLEOTIDES  \\\n",
       "0        NaN            1977.0           0.0               nan   \n",
       "1        1.0            4013.0           0.0               nan   \n",
       "2        0.0            7362.0           0.0               nan   \n",
       "\n",
       "                                        VCF_EVIDENCE  VCF_IDX  \\\n",
       "0  {\"GT\": [1, 1], \"DP\": 45.0, \"ALLELE_DP\": [0.0, ...      1.0   \n",
       "1  {\"GT\": [1, 1], \"DP\": 103.0, \"ALLELE_DP\": [0.0,...      1.0   \n",
       "2  {\"GT\": [1, 1], \"DP\": 61.0, \"ALLELE_DP\": [0.0, ...      1.0   \n",
       "\n",
       "   IS_MINOR_ALLELE MINOR_VARIANT  FRS  \n",
       "0            False          None  NaN  \n",
       "1            False          None  NaN  \n",
       "2            False          None  NaN  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VARIANTS = pandas.read_pickle('/Users/fowler/files/cryptic/cryptic-tables-v2.1.2/VARIANTS.pkl.gz')\n",
    "VARIANTS.reset_index(inplace=True)\n",
    "VARIANTS[:3]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MUTATION\n",
       "a-3g       89631\n",
       "A111A      83541\n",
       "L63L       80648\n",
       "A18A       77389\n",
       "A72A       71442\n",
       "I145M      65802\n",
       "A52A       64913\n",
       "S133S      61820\n",
       "R21R       60386\n",
       "A94A       57924\n",
       "A96A       57782\n",
       "Q68H       57712\n",
       "del_1.0    56643\n",
       "R180G      56503\n",
       "V210V      55975\n",
       "A32A       55618\n",
       "V235V      55091\n",
       "Q20L       54801\n",
       "A266A      54601\n",
       "c-61t      52951\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS.MUTATION.value_counts()[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['UNIQUEID', 'GENE', 'MUTATION', 'REF', 'ALT', 'NUCLEOTIDE_NUMBER',\n",
       "       'NUCLEOTIDE_INDEX', 'GENE_POSITION', 'CODES_PROTEIN', 'INDEL_LENGTH',\n",
       "       'INDEL_NUCLEOTIDES', 'AMINO_ACID_NUMBER', 'AMINO_ACID_SEQUENCE',\n",
       "       'NUMBER_NUCLEOTIDE_CHANGES', 'IS_MINOR_ALLELE', 'MINOR_MUTATION',\n",
       "       'FRS'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>NUCLEOTIDE_INDEX</th>\n",
       "      <th>GENE_POSITION</th>\n",
       "      <th>CODES_PROTEIN</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_NUCLEOTIDES</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>AMINO_ACID_SEQUENCE</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>MINOR_MUTATION</th>\n",
       "      <th>FRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>121146</th>\n",
       "      <td>site.01.subj.DR0623.lab.DR0623.iso.1</td>\n",
       "      <td>Rv0678</td>\n",
       "      <td>135_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>135.0</td>\n",
       "      <td>779124.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "      <td>tgtga</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>135_ins_tgtga</td>\n",
       "      <td>0.310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121148</th>\n",
       "      <td>site.01.subj.DR0623.lab.DR0623.iso.1</td>\n",
       "      <td>Rv0678</td>\n",
       "      <td>192_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>192.0</td>\n",
       "      <td>779181.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>g</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>192_ins_g</td>\n",
       "      <td>0.218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144304</th>\n",
       "      <td>site.01.subj.DR0144.lab.DR0144.iso.1</td>\n",
       "      <td>gid</td>\n",
       "      <td>352_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>352.0</td>\n",
       "      <td>4407851.0</td>\n",
       "      <td>352.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>g</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>352_ins_g</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207468</th>\n",
       "      <td>site.16.subj.001.lab.10398.iso.1</td>\n",
       "      <td>pncA</td>\n",
       "      <td>457_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>457.0</td>\n",
       "      <td>2288785.0</td>\n",
       "      <td>457.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>457_ins_c</td>\n",
       "      <td>0.851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210143</th>\n",
       "      <td>site.16.subj.001.lab.15323.iso.1</td>\n",
       "      <td>inhA</td>\n",
       "      <td>-10_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>1674192.0</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>False</td>\n",
       "      <td>11.0</td>\n",
       "      <td>taaggatattg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>-10_ins_taaggatattg</td>\n",
       "      <td>0.663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57561611</th>\n",
       "      <td>site.07.subj.C32A080B-63AC-4427-9257-073281798...</td>\n",
       "      <td>rrs</td>\n",
       "      <td>459_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>459.0</td>\n",
       "      <td>1472304.0</td>\n",
       "      <td>459.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>aca</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>459_ins_aca</td>\n",
       "      <td>0.727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57561612</th>\n",
       "      <td>site.07.subj.C32A080B-63AC-4427-9257-073281798...</td>\n",
       "      <td>rrs</td>\n",
       "      <td>461_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>461.0</td>\n",
       "      <td>1472306.0</td>\n",
       "      <td>461.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>gg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>461_ins_gg</td>\n",
       "      <td>0.714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57561616</th>\n",
       "      <td>site.07.subj.C32A080B-63AC-4427-9257-073281798...</td>\n",
       "      <td>rrs</td>\n",
       "      <td>483_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>483.0</td>\n",
       "      <td>1472328.0</td>\n",
       "      <td>483.0</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>gt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>483_ins_gt</td>\n",
       "      <td>0.632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57561618</th>\n",
       "      <td>site.07.subj.C32A080B-63AC-4427-9257-073281798...</td>\n",
       "      <td>rrs</td>\n",
       "      <td>589_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>589.0</td>\n",
       "      <td>1472434.0</td>\n",
       "      <td>589.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>589_ins_c</td>\n",
       "      <td>0.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57561620</th>\n",
       "      <td>site.07.subj.C32A080B-63AC-4427-9257-073281798...</td>\n",
       "      <td>rrs</td>\n",
       "      <td>601_minorindel</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>601.0</td>\n",
       "      <td>1472446.0</td>\n",
       "      <td>601.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>601_ins_c</td>\n",
       "      <td>0.429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10322 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   UNIQUEID    GENE  \\\n",
       "121146                 site.01.subj.DR0623.lab.DR0623.iso.1  Rv0678   \n",
       "121148                 site.01.subj.DR0623.lab.DR0623.iso.1  Rv0678   \n",
       "144304                 site.01.subj.DR0144.lab.DR0144.iso.1     gid   \n",
       "207468                     site.16.subj.001.lab.10398.iso.1    pncA   \n",
       "210143                     site.16.subj.001.lab.15323.iso.1    inhA   \n",
       "...                                                     ...     ...   \n",
       "57561611  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   \n",
       "57561612  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   \n",
       "57561616  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   \n",
       "57561618  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   \n",
       "57561620  site.07.subj.C32A080B-63AC-4427-9257-073281798...     rrs   \n",
       "\n",
       "                MUTATION  REF  ALT  NUCLEOTIDE_NUMBER  NUCLEOTIDE_INDEX  \\\n",
       "121146    135_minorindel  nan  nan              135.0          779124.0   \n",
       "121148    192_minorindel  nan  nan              192.0          779181.0   \n",
       "144304    352_minorindel  nan  nan              352.0         4407851.0   \n",
       "207468    457_minorindel  nan  nan              457.0         2288785.0   \n",
       "210143    -10_minorindel  nan  nan              -10.0         1674192.0   \n",
       "...                  ...  ...  ...                ...               ...   \n",
       "57561611  459_minorindel  nan  nan              459.0         1472304.0   \n",
       "57561612  461_minorindel  nan  nan              461.0         1472306.0   \n",
       "57561616  483_minorindel  nan  nan              483.0         1472328.0   \n",
       "57561618  589_minorindel  nan  nan              589.0         1472434.0   \n",
       "57561620  601_minorindel  nan  nan              601.0         1472446.0   \n",
       "\n",
       "          GENE_POSITION  CODES_PROTEIN  INDEL_LENGTH INDEL_NUCLEOTIDES  \\\n",
       "121146            135.0           True           5.0             tgtga   \n",
       "121148            192.0           True           1.0                 g   \n",
       "144304            352.0           True           1.0                 g   \n",
       "207468            457.0           True           1.0                 c   \n",
       "210143            -10.0          False          11.0       taaggatattg   \n",
       "...                 ...            ...           ...               ...   \n",
       "57561611          459.0          False           3.0               aca   \n",
       "57561612          461.0          False           2.0                gg   \n",
       "57561616          483.0          False           2.0                gt   \n",
       "57561618          589.0          False           1.0                 c   \n",
       "57561620          601.0          False           1.0                 c   \n",
       "\n",
       "          AMINO_ACID_NUMBER AMINO_ACID_SEQUENCE  NUMBER_NUCLEOTIDE_CHANGES  \\\n",
       "121146                  NaN                 nan                        0.0   \n",
       "121148                  NaN                 nan                        0.0   \n",
       "144304                  NaN                 nan                        0.0   \n",
       "207468                  NaN                 nan                        0.0   \n",
       "210143                  NaN                 nan                        0.0   \n",
       "...                     ...                 ...                        ...   \n",
       "57561611                NaN                 nan                        0.0   \n",
       "57561612                NaN                 nan                        0.0   \n",
       "57561616                NaN                 nan                        0.0   \n",
       "57561618                NaN                 nan                        0.0   \n",
       "57561620                NaN                 nan                        0.0   \n",
       "\n",
       "          IS_MINOR_ALLELE       MINOR_MUTATION    FRS  \n",
       "121146               True        135_ins_tgtga  0.310  \n",
       "121148               True            192_ins_g  0.218  \n",
       "144304               True            352_ins_g  0.250  \n",
       "207468               True            457_ins_c  0.851  \n",
       "210143               True  -10_ins_taaggatattg  0.663  \n",
       "...                   ...                  ...    ...  \n",
       "57561611             True          459_ins_aca  0.727  \n",
       "57561612             True           461_ins_gg  0.714  \n",
       "57561616             True           483_ins_gt  0.632  \n",
       "57561618             True            589_ins_c  0.400  \n",
       "57561620             True            601_ins_c  0.429  \n",
       "\n",
       "[10322 rows x 17 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS[MUTATIONS.IS_MINOR_ALLELE & (MUTATIONS.INDEL_LENGTH>0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>NUCLEOTIDE_NUMBER</th>\n",
       "      <th>NUCLEOTIDE_INDEX</th>\n",
       "      <th>GENE_POSITION</th>\n",
       "      <th>CODES_PROTEIN</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_NUCLEOTIDES</th>\n",
       "      <th>AMINO_ACID_NUMBER</th>\n",
       "      <th>AMINO_ACID_SEQUENCE</th>\n",
       "      <th>NUMBER_NUCLEOTIDE_CHANGES</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>MINOR_MUTATION</th>\n",
       "      <th>FRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [UNIQUEID, GENE, MUTATION, REF, ALT, NUCLEOTIDE_NUMBER, NUCLEOTIDE_INDEX, GENE_POSITION, CODES_PROTEIN, INDEL_LENGTH, INDEL_NUCLEOTIDES, AMINO_ACID_NUMBER, AMINO_ACID_SEQUENCE, NUMBER_NUCLEOTIDE_CHANGES, IS_MINOR_ALLELE, MINOR_MUTATION, FRS]\n",
       "Index: []"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MUTATIONS[MUTATIONS.MUTATION.str[-1]=='x']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>gene</th>\n",
       "      <th>mutation</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>nucleotide_number</th>\n",
       "      <th>nucleotide_index</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>codes_protein</th>\n",
       "      <th>indel_length</th>\n",
       "      <th>indel_nucleotides</th>\n",
       "      <th>amino_acid_number</th>\n",
       "      <th>amino_acid_sequence</th>\n",
       "      <th>number_nucleotide_changes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>A511X</td>\n",
       "      <td>gcc</td>\n",
       "      <td>gxc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>511.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>511.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>A513X</td>\n",
       "      <td>gcg</td>\n",
       "      <td>gxg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>513.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>513.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>L515X</td>\n",
       "      <td>ttg</td>\n",
       "      <td>ttx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>515.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>515.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>A517X</td>\n",
       "      <td>gcc</td>\n",
       "      <td>xcc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>517.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>517.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS23</td>\n",
       "      <td>G443X</td>\n",
       "      <td>ggc</td>\n",
       "      <td>ggx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>443.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>443.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS4</td>\n",
       "      <td>!838X</td>\n",
       "      <td>taa</td>\n",
       "      <td>tax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>838.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>838.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS47</td>\n",
       "      <td>S271X</td>\n",
       "      <td>agc</td>\n",
       "      <td>xgc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>271.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>271.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS54</td>\n",
       "      <td>G594X</td>\n",
       "      <td>ggg</td>\n",
       "      <td>ggx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>594.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>594.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS56</td>\n",
       "      <td>D177X</td>\n",
       "      <td>gat</td>\n",
       "      <td>gxt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>177.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>177.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS57</td>\n",
       "      <td>G265X</td>\n",
       "      <td>ggc</td>\n",
       "      <td>ggx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>265.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>265.0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   uniqueid       gene mutation  ref  alt  \\\n",
       "16   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A511X  gcc  gxc   \n",
       "17   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A513X  gcg  gxg   \n",
       "18   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    L515X  ttg  ttx   \n",
       "19   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A517X  gcc  xcc   \n",
       "62   site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS23    G443X  ggc  ggx   \n",
       "112  site.ENA.subj.SAMN07766182.lab.1.iso.1   PE_PGRS4    !838X  taa  tax   \n",
       "128  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS47    S271X  agc  xgc   \n",
       "144  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS54    G594X  ggg  ggx   \n",
       "162  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS56    D177X  gat  gxt   \n",
       "163  site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS57    G265X  ggc  ggx   \n",
       "\n",
       "     nucleotide_number  nucleotide_index  gene_position  codes_protein  \\\n",
       "16                 NaN               NaN          511.0           True   \n",
       "17                 NaN               NaN          513.0           True   \n",
       "18                 NaN               NaN          515.0           True   \n",
       "19                 NaN               NaN          517.0           True   \n",
       "62                 NaN               NaN          443.0           True   \n",
       "112                NaN               NaN          838.0           True   \n",
       "128                NaN               NaN          271.0           True   \n",
       "144                NaN               NaN          594.0           True   \n",
       "162                NaN               NaN          177.0           True   \n",
       "163                NaN               NaN          265.0           True   \n",
       "\n",
       "     indel_length indel_nucleotides  amino_acid_number amino_acid_sequence  \\\n",
       "16            NaN              None              511.0                   X   \n",
       "17            NaN              None              513.0                   X   \n",
       "18            NaN              None              515.0                   X   \n",
       "19            NaN              None              517.0                   X   \n",
       "62            NaN              None              443.0                   X   \n",
       "112           NaN              None              838.0                   X   \n",
       "128           NaN              None              271.0                   X   \n",
       "144           NaN              None              594.0                   X   \n",
       "162           NaN              None              177.0                   X   \n",
       "163           NaN              None              265.0                   X   \n",
       "\n",
       "     number_nucleotide_changes  \n",
       "16                         1.0  \n",
       "17                         1.0  \n",
       "18                         1.0  \n",
       "19                         1.0  \n",
       "62                         1.0  \n",
       "112                        1.0  \n",
       "128                        1.0  \n",
       "144                        1.0  \n",
       "162                        1.0  \n",
       "163                        1.0  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations[mutations.mutation.str.contains('X')][:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>gene</th>\n",
       "      <th>mutation</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>nucleotide_number</th>\n",
       "      <th>nucleotide_index</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>codes_protein</th>\n",
       "      <th>indel_length</th>\n",
       "      <th>indel_nucleotides</th>\n",
       "      <th>amino_acid_number</th>\n",
       "      <th>amino_acid_sequence</th>\n",
       "      <th>number_nucleotide_changes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE35</td>\n",
       "      <td>E99!:15</td>\n",
       "      <td>gaa</td>\n",
       "      <td>taa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>99.0</td>\n",
       "      <td>!</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>A180A:10</td>\n",
       "      <td>gcc</td>\n",
       "      <td>gcg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>180.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>180.0</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>R225G:22</td>\n",
       "      <td>agg</td>\n",
       "      <td>ggg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>225.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>225.0</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>G509G:3</td>\n",
       "      <td>ggc</td>\n",
       "      <td>ggt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>509.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>509.0</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>site.ENA.subj.SAMN07766182.lab.1.iso.1</td>\n",
       "      <td>PE_PGRS10</td>\n",
       "      <td>A518G:4</td>\n",
       "      <td>gcc</td>\n",
       "      <td>ggc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>518.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>518.0</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2511</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>T794I:6</td>\n",
       "      <td>acc</td>\n",
       "      <td>atc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>794.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>794.0</td>\n",
       "      <td>I</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2684</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>pks12</td>\n",
       "      <td>A3586V:46</td>\n",
       "      <td>gcg</td>\n",
       "      <td>gtg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3586.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>3586.0</td>\n",
       "      <td>V</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2685</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>pks12</td>\n",
       "      <td>G3602G:60</td>\n",
       "      <td>ggt</td>\n",
       "      <td>ggc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3602.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>3602.0</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2720</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>plcC</td>\n",
       "      <td>Y97N:3</td>\n",
       "      <td>tac</td>\n",
       "      <td>aac</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>97.0</td>\n",
       "      <td>N</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2854</th>\n",
       "      <td>site.ENA.subj.SAMEA104394448.lab.1.iso.1</td>\n",
       "      <td>rpfA</td>\n",
       "      <td>A209A:4</td>\n",
       "      <td>gca</td>\n",
       "      <td>gcg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>209.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>209.0</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17453 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      uniqueid       gene   mutation  ref  \\\n",
       "9       site.ENA.subj.SAMN07766182.lab.1.iso.1       PE35    E99!:15  gaa   \n",
       "21      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10   A180A:10  gcc   \n",
       "22      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10   R225G:22  agg   \n",
       "23      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    G509G:3  ggc   \n",
       "24      site.ENA.subj.SAMN07766182.lab.1.iso.1  PE_PGRS10    A518G:4  gcc   \n",
       "...                                        ...        ...        ...  ...   \n",
       "2511  site.ENA.subj.SAMEA104394448.lab.1.iso.1      mmpL5    T794I:6  acc   \n",
       "2684  site.ENA.subj.SAMEA104394448.lab.1.iso.1      pks12  A3586V:46  gcg   \n",
       "2685  site.ENA.subj.SAMEA104394448.lab.1.iso.1      pks12  G3602G:60  ggt   \n",
       "2720  site.ENA.subj.SAMEA104394448.lab.1.iso.1       plcC     Y97N:3  tac   \n",
       "2854  site.ENA.subj.SAMEA104394448.lab.1.iso.1       rpfA    A209A:4  gca   \n",
       "\n",
       "      alt  nucleotide_number  nucleotide_index  gene_position  codes_protein  \\\n",
       "9     taa                NaN               NaN           99.0           True   \n",
       "21    gcg                NaN               NaN          180.0           True   \n",
       "22    ggg                NaN               NaN          225.0           True   \n",
       "23    ggt                NaN               NaN          509.0           True   \n",
       "24    ggc                NaN               NaN          518.0           True   \n",
       "...   ...                ...               ...            ...            ...   \n",
       "2511  atc                NaN               NaN          794.0           True   \n",
       "2684  gtg                NaN               NaN         3586.0           True   \n",
       "2685  ggc                NaN               NaN         3602.0           True   \n",
       "2720  aac                NaN               NaN           97.0           True   \n",
       "2854  gcg                NaN               NaN          209.0           True   \n",
       "\n",
       "      indel_length indel_nucleotides  amino_acid_number amino_acid_sequence  \\\n",
       "9              NaN              None               99.0                   !   \n",
       "21             NaN              None              180.0                   A   \n",
       "22             NaN              None              225.0                   G   \n",
       "23             NaN              None              509.0                   G   \n",
       "24             NaN              None              518.0                   G   \n",
       "...            ...               ...                ...                 ...   \n",
       "2511           NaN              None              794.0                   I   \n",
       "2684           NaN              None             3586.0                   V   \n",
       "2685           NaN              None             3602.0                   G   \n",
       "2720           NaN              None               97.0                   N   \n",
       "2854           NaN              None              209.0                   A   \n",
       "\n",
       "      number_nucleotide_changes  \n",
       "9                           1.0  \n",
       "21                          1.0  \n",
       "22                          1.0  \n",
       "23                          1.0  \n",
       "24                          1.0  \n",
       "...                         ...  \n",
       "2511                        1.0  \n",
       "2684                        1.0  \n",
       "2685                        1.0  \n",
       "2720                        1.0  \n",
       "2854                        1.0  \n",
       "\n",
       "[17453 rows x 14 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations[mutations.mutation.str.contains(':')] # & (mutations.indel_length==0)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mutation -> mutation, is_null, is_minor_allele, minor_mutation, minor_reads\n",
    "\n",
    "`1241_ins_cgccggcgg:10` parsed to `1241_minorindel`, `False`, `True`, `1241_ins_cgccggcgg`, `10`\n",
    "\n",
    "`E99!:15` parsed to `E99Z`, `False`, `True`, `E99!`, `15`\n",
    "\n",
    "`g-57x` parsed to `g-57x`, `True`, `False`, None, None\n",
    "\n",
    "`A511X` parsed to `g-A511X`, `True`, `False`, None, None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation \n",
    "\n",
    "def parse_mutations(row):\n",
    "    if ':' in row.mutation:\n",
    "        if indel_length > 0 or indel_length < 0:\n",
    "            mut\n",
    "        mut,minor_reads = row.mutation.split(':')\n",
    "        is_minor = True\n",
    "    else:\n",
    "        mut = row.mutation\n",
    "        minor_reads = None\n",
    "        is_minor = False\n",
    "\n",
    "    return [row.mutation]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>VARIANT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>GENE_POSITION</th>\n",
       "      <th>CODON_IDX</th>\n",
       "      <th>NUCLEOTIDE_INDEX</th>\n",
       "      <th>INDEL_LENGTH</th>\n",
       "      <th>INDEL_NUCLEOTIDES</th>\n",
       "      <th>VCF_EVIDENCE</th>\n",
       "      <th>VCF_IDX</th>\n",
       "      <th>IS_MINOR_ALLELE</th>\n",
       "      <th>MINOR_VARIANT</th>\n",
       "      <th>FRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7950</th>\n",
       "      <td>site.01.subj.DR0611.lab.DR0611.iso.1</td>\n",
       "      <td>760729a&gt;z</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>308.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>760729.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [1, 1], \"DP\": 77.0, \"ALLELE_DP\": [12.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>760729a&gt;g</td>\n",
       "      <td>0.865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16139</th>\n",
       "      <td>site.01.subj.DR0600.lab.DR0600.iso.1</td>\n",
       "      <td>1472805c&gt;z</td>\n",
       "      <td>rrs</td>\n",
       "      <td>960.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1472805.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [0, 0], \"DP\": 47.0, \"ALLELE_DP\": [47.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1472805c&gt;t</td>\n",
       "      <td>0.230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34546</th>\n",
       "      <td>site.01.subj.DR0685.lab.DR0685.iso.1</td>\n",
       "      <td>7572t&gt;z</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>91.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7572.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [0, 0], \"DP\": 36.0, \"ALLELE_DP\": [36.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>7572t&gt;c</td>\n",
       "      <td>0.357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34547</th>\n",
       "      <td>site.01.subj.DR0685.lab.DR0685.iso.1</td>\n",
       "      <td>7582a&gt;z</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7582.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [1, 1], \"DP\": 39.0, \"ALLELE_DP\": [24.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>7582a&gt;g</td>\n",
       "      <td>0.619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41351</th>\n",
       "      <td>site.01.subj.DR0653.lab.DR0653.iso.1</td>\n",
       "      <td>8164c&gt;z</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>288.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8164.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [1, 1], \"DP\": 121.0, \"ALLELE_DP\": [17.0...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>8164c&gt;a</td>\n",
       "      <td>0.877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79166</th>\n",
       "      <td>site.01.subj.DR0589.lab.DR0589.iso.1</td>\n",
       "      <td>761100c&gt;z</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>432.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>761100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [0, 0], \"DP\": 37.0, \"ALLELE_DP\": [37.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>761100c&gt;a</td>\n",
       "      <td>0.098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79167</th>\n",
       "      <td>site.01.subj.DR0589.lab.DR0589.iso.1</td>\n",
       "      <td>761139c&gt;z</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>445.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>761139.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [0, 0], \"DP\": 33.0, \"ALLELE_DP\": [33.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>761139c&gt;a</td>\n",
       "      <td>0.283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79168</th>\n",
       "      <td>site.01.subj.DR0589.lab.DR0589.iso.1</td>\n",
       "      <td>761161t&gt;z</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>452.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>761161.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [1, 1], \"DP\": 32.0, \"ALLELE_DP\": [18.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>761161t&gt;c</td>\n",
       "      <td>0.640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93493</th>\n",
       "      <td>site.01.subj.DR0679.lab.DR0679.iso.1</td>\n",
       "      <td>7581g&gt;z</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7581.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [1, 1], \"DP\": 43.0, \"ALLELE_DP\": [25.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>7581g&gt;a</td>\n",
       "      <td>0.632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93494</th>\n",
       "      <td>site.01.subj.DR0679.lab.DR0679.iso.1</td>\n",
       "      <td>7582a&gt;z</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>94.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7582.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>{\"GT\": [0, 0], \"DP\": 46.0, \"ALLELE_DP\": [46.0,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>7582a&gt;g</td>\n",
       "      <td>0.352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   UNIQUEID     VARIANT  GENE  GENE_POSITION  \\\n",
       "7950   site.01.subj.DR0611.lab.DR0611.iso.1   760729a>z  rpoB          308.0   \n",
       "16139  site.01.subj.DR0600.lab.DR0600.iso.1  1472805c>z   rrs          960.0   \n",
       "34546  site.01.subj.DR0685.lab.DR0685.iso.1     7572t>z  gyrA           91.0   \n",
       "34547  site.01.subj.DR0685.lab.DR0685.iso.1     7582a>z  gyrA           94.0   \n",
       "41351  site.01.subj.DR0653.lab.DR0653.iso.1     8164c>z  gyrA          288.0   \n",
       "79166  site.01.subj.DR0589.lab.DR0589.iso.1   761100c>z  rpoB          432.0   \n",
       "79167  site.01.subj.DR0589.lab.DR0589.iso.1   761139c>z  rpoB          445.0   \n",
       "79168  site.01.subj.DR0589.lab.DR0589.iso.1   761161t>z  rpoB          452.0   \n",
       "93493  site.01.subj.DR0679.lab.DR0679.iso.1     7581g>z  gyrA           94.0   \n",
       "93494  site.01.subj.DR0679.lab.DR0679.iso.1     7582a>z  gyrA           94.0   \n",
       "\n",
       "       CODON_IDX  NUCLEOTIDE_INDEX  INDEL_LENGTH INDEL_NUCLEOTIDES  \\\n",
       "7950         2.0          760729.0           0.0               nan   \n",
       "16139        NaN         1472805.0           0.0               nan   \n",
       "34546        1.0            7572.0           0.0               nan   \n",
       "34547        2.0            7582.0           0.0               nan   \n",
       "41351        2.0            8164.0           0.0               nan   \n",
       "79166        1.0          761100.0           0.0               nan   \n",
       "79167        1.0          761139.0           0.0               nan   \n",
       "79168        2.0          761161.0           0.0               nan   \n",
       "93493        1.0            7581.0           0.0               nan   \n",
       "93494        2.0            7582.0           0.0               nan   \n",
       "\n",
       "                                            VCF_EVIDENCE  VCF_IDX  \\\n",
       "7950   {\"GT\": [1, 1], \"DP\": 77.0, \"ALLELE_DP\": [12.0,...      1.0   \n",
       "16139  {\"GT\": [0, 0], \"DP\": 47.0, \"ALLELE_DP\": [47.0,...      1.0   \n",
       "34546  {\"GT\": [0, 0], \"DP\": 36.0, \"ALLELE_DP\": [36.0,...      1.0   \n",
       "34547  {\"GT\": [1, 1], \"DP\": 39.0, \"ALLELE_DP\": [24.0,...      1.0   \n",
       "41351  {\"GT\": [1, 1], \"DP\": 121.0, \"ALLELE_DP\": [17.0...      1.0   \n",
       "79166  {\"GT\": [0, 0], \"DP\": 37.0, \"ALLELE_DP\": [37.0,...      1.0   \n",
       "79167  {\"GT\": [0, 0], \"DP\": 33.0, \"ALLELE_DP\": [33.0,...      1.0   \n",
       "79168  {\"GT\": [1, 1], \"DP\": 32.0, \"ALLELE_DP\": [18.0,...      1.0   \n",
       "93493  {\"GT\": [1, 1], \"DP\": 43.0, \"ALLELE_DP\": [25.0,...      1.0   \n",
       "93494  {\"GT\": [0, 0], \"DP\": 46.0, \"ALLELE_DP\": [46.0,...      1.0   \n",
       "\n",
       "       IS_MINOR_ALLELE MINOR_VARIANT    FRS  \n",
       "7950              True     760729a>g  0.865  \n",
       "16139             True    1472805c>t  0.230  \n",
       "34546             True       7572t>c  0.357  \n",
       "34547             True       7582a>g  0.619  \n",
       "41351             True       8164c>a  0.877  \n",
       "79166             True     761100c>a  0.098  \n",
       "79167             True     761139c>a  0.283  \n",
       "79168             True     761161t>c  0.640  \n",
       "93493             True       7581g>a  0.632  \n",
       "93494             True       7582a>g  0.352  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VARIANTS[VARIANTS.IS_MINOR_ALLELE][:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>codon_idx</th>\n",
       "      <th>nucleotide_index</th>\n",
       "      <th>indel_length</th>\n",
       "      <th>indel_nucleotides</th>\n",
       "      <th>vcf_evidence</th>\n",
       "      <th>vcf_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>79571c&gt;x</td>\n",
       "      <td>Rv0071</td>\n",
       "      <td>29.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>79571</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"ALLELE_DP\": [0, 0], \"FRS\": \".\", \"GT_CONF_PER...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>80616c&gt;x</td>\n",
       "      <td>Rv0072</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80616</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT\": [1, 1], \"FRS\": 1.0, \"ALLELE_DP\": [0, 1]...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>336560t&gt;x</td>\n",
       "      <td>PE_PGRS4</td>\n",
       "      <td>838.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>336560</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT\": [0, 0], \"COV_TOTAL\": 1, \"GT_CONF\": 7.69...</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>336590g&gt;x</td>\n",
       "      <td>PE_PGRS4</td>\n",
       "      <td>828.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>336590</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"FRS\": 0.5, \"GT\": [0, 0], \"GT_CONF\": 0.0, \"CO...</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>336592t&gt;x</td>\n",
       "      <td>PE_PGRS4</td>\n",
       "      <td>828.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>336592</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"FRS\": 0.5, \"COV\": [1, 1], \"COV_TOTAL\": 2, \"G...</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>3739498c&gt;x</td>\n",
       "      <td>PE_PGRS50</td>\n",
       "      <td>1093.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3739498</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT_CONF\": 22.3, \"ALLELE_DP\": [0, 4], \"DP\": 4...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>3934022t&gt;x</td>\n",
       "      <td>PE_PGRS54</td>\n",
       "      <td>1006.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3934022</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT\": [1, 1], \"GT_CONF_PERCENTILE\": 0.04, \"CO...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>3935950t&gt;x</td>\n",
       "      <td>PE_PGRS54</td>\n",
       "      <td>1649.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3935950</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV_TOTAL\": 0, \"COV\": [0, 0], \"GT_CONF\": 0.0...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>3946813a&gt;x</td>\n",
       "      <td>PE_PGRS57</td>\n",
       "      <td>340.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3946813</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT_CONF_PERCENTILE\": 0.04, \"COV\": [0, 1], \"A...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>3948612a&gt;x</td>\n",
       "      <td>PE_PGRS57</td>\n",
       "      <td>940.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3948612</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV_TOTAL\": 3, \"GT_CONF\": 16.0, \"DP\": 2, \"GT...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6083 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   uniqueid     variant       gene  \\\n",
       "45   site.ENA.subj.SAMN01908825.lab.1.iso.1    79571c>x     Rv0071   \n",
       "46   site.ENA.subj.SAMN01908825.lab.1.iso.1    80616c>x     Rv0072   \n",
       "144  site.ENA.subj.SAMN01908825.lab.1.iso.1   336560t>x   PE_PGRS4   \n",
       "145  site.ENA.subj.SAMN01908825.lab.1.iso.1   336590g>x   PE_PGRS4   \n",
       "146  site.ENA.subj.SAMN01908825.lab.1.iso.1   336592t>x   PE_PGRS4   \n",
       "..                                      ...         ...        ...   \n",
       "721  site.ENA.subj.SAMN07659930.lab.1.iso.1  3739498c>x  PE_PGRS50   \n",
       "762  site.ENA.subj.SAMN07659930.lab.1.iso.1  3934022t>x  PE_PGRS54   \n",
       "764  site.ENA.subj.SAMN07659930.lab.1.iso.1  3935950t>x  PE_PGRS54   \n",
       "765  site.ENA.subj.SAMN07659930.lab.1.iso.1  3946813a>x  PE_PGRS57   \n",
       "766  site.ENA.subj.SAMN07659930.lab.1.iso.1  3948612a>x  PE_PGRS57   \n",
       "\n",
       "     gene_position  codon_idx  nucleotide_index  indel_length  \\\n",
       "45            29.0        1.0             79571             0   \n",
       "46            -8.0        NaN             80616             0   \n",
       "144          838.0        2.0            336560             0   \n",
       "145          828.0        2.0            336590             0   \n",
       "146          828.0        0.0            336592             0   \n",
       "..             ...        ...               ...           ...   \n",
       "721         1093.0        0.0           3739498             0   \n",
       "762         1006.0        2.0           3934022             0   \n",
       "764         1649.0        1.0           3935950             0   \n",
       "765          340.0        2.0           3946813             0   \n",
       "766          940.0        1.0           3948612             0   \n",
       "\n",
       "    indel_nucleotides                                       vcf_evidence  \\\n",
       "45               None  {\"ALLELE_DP\": [0, 0], \"FRS\": \".\", \"GT_CONF_PER...   \n",
       "46               None  {\"GT\": [1, 1], \"FRS\": 1.0, \"ALLELE_DP\": [0, 1]...   \n",
       "144              None  {\"GT\": [0, 0], \"COV_TOTAL\": 1, \"GT_CONF\": 7.69...   \n",
       "145              None  {\"FRS\": 0.5, \"GT\": [0, 0], \"GT_CONF\": 0.0, \"CO...   \n",
       "146              None  {\"FRS\": 0.5, \"COV\": [1, 1], \"COV_TOTAL\": 2, \"G...   \n",
       "..                ...                                                ...   \n",
       "721              None  {\"GT_CONF\": 22.3, \"ALLELE_DP\": [0, 4], \"DP\": 4...   \n",
       "762              None  {\"GT\": [1, 1], \"GT_CONF_PERCENTILE\": 0.04, \"CO...   \n",
       "764              None  {\"COV_TOTAL\": 0, \"COV\": [0, 0], \"GT_CONF\": 0.0...   \n",
       "765              None  {\"GT_CONF_PERCENTILE\": 0.04, \"COV\": [0, 1], \"A...   \n",
       "766              None  {\"COV_TOTAL\": 3, \"GT_CONF\": 16.0, \"DP\": 2, \"GT...   \n",
       "\n",
       "     vcf_idx  \n",
       "45       NaN  \n",
       "46       1.0  \n",
       "144      0.0  \n",
       "145      0.0  \n",
       "146      0.0  \n",
       "..       ...  \n",
       "721      1.0  \n",
       "762      1.0  \n",
       "764      NaN  \n",
       "765      1.0  \n",
       "766      1.0  \n",
       "\n",
       "[6083 rows x 10 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants[variants.variant.str.contains('x')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "variant -> variant, is_null, is_minor_allele, minor_varian, minor_reads, total_reads\n",
    "\n",
    "`7570c>t:76` parsed to `7570c>z`, `False`, `True`, `7570c>t`, 76, 110\n",
    "\n",
    "`24699_del_ccgcgttgctcggggtaa:41` parsed to `24699_minorindel`, `False`, `True`, `24699_del_ccgcgttgctcggggtaa`, 41, 47\n",
    "\n",
    "`338554t>x` parsed to `338554t>x`, `True`, `False`, None, None\t\t\t\n",
    "\n",
    "\n",
    "```\n",
    "\"{\"\"COV\"\": [34, 76], \"\"COV_TOTAL\"\": 110, \"\"GT\"\": [1, 1], \"\"GT_CONF\"\": 261.86, \"\"GT_CONF_PERCENTILE\"\": 6.47, \"\"ALLELE_DP\"\": [34, 76], \"\"DP\"\": 76, \"\"FRS\"\": 0.6909, \"\"POS\"\": 7570, \"\"REF\"\": \"\"c\"\", \"\"ALTS\"\": [\"\"t\"\"]}\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json \n",
    "\n",
    "def parse_variants(row):\n",
    "\n",
    "    variant = row.variant\n",
    "    is_null = False\n",
    "    is_minor = False\n",
    "    minor_variant = None\n",
    "    minor_reads = None\n",
    "    coverage = None\n",
    "    frs = None\n",
    "\n",
    "    a=(json.loads(row['vcf_evidence']))\n",
    "    if 'COV_TOTAL' in a.keys():\n",
    "        coverage = a['COV_TOTAL']\n",
    "    if 'FRS' in a.keys():\n",
    "        frs = a['FRS']\n",
    "\n",
    "    if row.variant[-1]=='x':\n",
    "        is_null = True\n",
    "    elif ':' in row.variant:\n",
    "        is_minor = True\n",
    "        minor_variant, minor_reads = row.variant.split(':')\n",
    "        if \"ins\" in variant or \"del\" in variant:\n",
    "            variant = minor_variant.split('_')[0]+\"_minorindel\"\n",
    "        else:\n",
    "            variant = minor_variant[:-1]+'x'\n",
    "\n",
    "    return pandas.Series([variant, is_null, is_minor, minor_variant, minor_reads, coverage, frs])\n",
    "\n",
    "variants[['var','is_null','is_minor','minor_variant','minor_reads','coverage','FRS']] = variants.apply(parse_variants, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>codon_idx</th>\n",
       "      <th>nucleotide_index</th>\n",
       "      <th>indel_length</th>\n",
       "      <th>indel_nucleotides</th>\n",
       "      <th>vcf_evidence</th>\n",
       "      <th>vcf_idx</th>\n",
       "      <th>var</th>\n",
       "      <th>is_null</th>\n",
       "      <th>is_minor</th>\n",
       "      <th>minor_variant</th>\n",
       "      <th>minor_reads</th>\n",
       "      <th>coverage</th>\n",
       "      <th>FRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>1977a&gt;g</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>-75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1977</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"DP\": 115, \"GT\": [1, 1], \"GT_CONF\": 738.27, \"...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1977a&gt;g</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>115.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>4013t&gt;c</td>\n",
       "      <td>recF</td>\n",
       "      <td>245.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4013</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"ALLELE_DP\": [0, 110], \"COV\": [0, 110], \"DP\":...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4013t&gt;c</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>7362g&gt;c</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7362</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"ALLELE_DP\": [0, 127], \"GT\": [1, 1], \"DP\": 12...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7362g&gt;c</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>127.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>7585g&gt;c</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>95.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7585</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"ALLELE_DP\": [0, 114], \"COV_TOTAL\": 114, \"GT\"...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7585g&gt;c</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>114.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>9304g&gt;a</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>668.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9304</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT_CONF_PERCENTILE\": 68.63, \"COV_TOTAL\": 114...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9304g&gt;a</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>114.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>916</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>3948649a&gt;c:4</td>\n",
       "      <td>PE_PGRS57</td>\n",
       "      <td>952.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3948649</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV_TOTAL\": 7, \"FRS\": 0.5714, \"GT\": [1, 1], ...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3948649a&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>3948649a&gt;c</td>\n",
       "      <td>4</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.5714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>917</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>4032006c&gt;g:4</td>\n",
       "      <td>PE_PGRS58</td>\n",
       "      <td>385.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4032006</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV\": [0, 4], \"GT\": [1, 1], \"DP\": 4, \"ALLELE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4032006c&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4032006c&gt;g</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>918</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>4036926_del_gccgcccgc:4</td>\n",
       "      <td>PE_PGRS59</td>\n",
       "      <td>1117.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4036926</td>\n",
       "      <td>-9</td>\n",
       "      <td>gccgcccgc</td>\n",
       "      <td>{\"GT_CONF\": 36.62, \"COV_TOTAL\": 4, \"GT_CONF_PE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4036926_minorindel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4036926_del_gccgcccgc</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>4060210t&gt;a:61</td>\n",
       "      <td>esxV</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4060210</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV\": [9, 61], \"FRS\": 0.8714, \"GT_CONF_PERCE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4060210t&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4060210t&gt;a</td>\n",
       "      <td>61</td>\n",
       "      <td>70.0</td>\n",
       "      <td>0.8714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>4263501t&gt;g:9</td>\n",
       "      <td>Rv3802c</td>\n",
       "      <td>289.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4263501</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT_CONF\": 311.55, \"GT_CONF_PERCENTILE\": 52.8...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4263501t&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4263501t&gt;g</td>\n",
       "      <td>9</td>\n",
       "      <td>68.0</td>\n",
       "      <td>0.8676</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120882 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   uniqueid                  variant  \\\n",
       "0    site.ENA.subj.SAMN01908825.lab.1.iso.1                  1977a>g   \n",
       "1    site.ENA.subj.SAMN01908825.lab.1.iso.1                  4013t>c   \n",
       "2    site.ENA.subj.SAMN01908825.lab.1.iso.1                  7362g>c   \n",
       "3    site.ENA.subj.SAMN01908825.lab.1.iso.1                  7585g>c   \n",
       "4    site.ENA.subj.SAMN01908825.lab.1.iso.1                  9304g>a   \n",
       "..                                      ...                      ...   \n",
       "916  site.ENA.subj.SAMN07659930.lab.1.iso.1             3948649a>c:4   \n",
       "917  site.ENA.subj.SAMN07659930.lab.1.iso.1             4032006c>g:4   \n",
       "918  site.ENA.subj.SAMN07659930.lab.1.iso.1  4036926_del_gccgcccgc:4   \n",
       "919  site.ENA.subj.SAMN07659930.lab.1.iso.1            4060210t>a:61   \n",
       "920  site.ENA.subj.SAMN07659930.lab.1.iso.1             4263501t>g:9   \n",
       "\n",
       "          gene  gene_position  codon_idx  nucleotide_index  indel_length  \\\n",
       "0         dnaN          -75.0        NaN              1977             0   \n",
       "1         recF          245.0        1.0              4013             0   \n",
       "2         gyrA           21.0        0.0              7362             0   \n",
       "3         gyrA           95.0        1.0              7585             0   \n",
       "4         gyrA          668.0        1.0              9304             0   \n",
       "..         ...            ...        ...               ...           ...   \n",
       "916  PE_PGRS57          952.0        2.0           3948649             0   \n",
       "917  PE_PGRS58          385.0        0.0           4032006             0   \n",
       "918  PE_PGRS59         1117.0        2.0           4036926            -9   \n",
       "919       esxV           20.0        1.0           4060210             0   \n",
       "920    Rv3802c          289.0        0.0           4263501             0   \n",
       "\n",
       "    indel_nucleotides                                       vcf_evidence  \\\n",
       "0                None  {\"DP\": 115, \"GT\": [1, 1], \"GT_CONF\": 738.27, \"...   \n",
       "1                None  {\"ALLELE_DP\": [0, 110], \"COV\": [0, 110], \"DP\":...   \n",
       "2                None  {\"ALLELE_DP\": [0, 127], \"GT\": [1, 1], \"DP\": 12...   \n",
       "3                None  {\"ALLELE_DP\": [0, 114], \"COV_TOTAL\": 114, \"GT\"...   \n",
       "4                None  {\"GT_CONF_PERCENTILE\": 68.63, \"COV_TOTAL\": 114...   \n",
       "..                ...                                                ...   \n",
       "916              None  {\"COV_TOTAL\": 7, \"FRS\": 0.5714, \"GT\": [1, 1], ...   \n",
       "917              None  {\"COV\": [0, 4], \"GT\": [1, 1], \"DP\": 4, \"ALLELE...   \n",
       "918         gccgcccgc  {\"GT_CONF\": 36.62, \"COV_TOTAL\": 4, \"GT_CONF_PE...   \n",
       "919              None  {\"COV\": [9, 61], \"FRS\": 0.8714, \"GT_CONF_PERCE...   \n",
       "920              None  {\"GT_CONF\": 311.55, \"GT_CONF_PERCENTILE\": 52.8...   \n",
       "\n",
       "     vcf_idx                 var  is_null  is_minor          minor_variant  \\\n",
       "0        1.0             1977a>g    False     False                   None   \n",
       "1        1.0             4013t>c    False     False                   None   \n",
       "2        1.0             7362g>c    False     False                   None   \n",
       "3        1.0             7585g>c    False     False                   None   \n",
       "4        1.0             9304g>a    False     False                   None   \n",
       "..       ...                 ...      ...       ...                    ...   \n",
       "916      1.0          3948649a>x    False      True             3948649a>c   \n",
       "917      1.0          4032006c>x    False      True             4032006c>g   \n",
       "918      1.0  4036926_minorindel    False      True  4036926_del_gccgcccgc   \n",
       "919      1.0          4060210t>x    False      True             4060210t>a   \n",
       "920      1.0          4263501t>x    False      True             4263501t>g   \n",
       "\n",
       "    minor_reads  coverage     FRS  \n",
       "0          None     115.0     1.0  \n",
       "1          None     110.0     1.0  \n",
       "2          None     127.0     1.0  \n",
       "3          None     114.0     1.0  \n",
       "4          None     114.0     1.0  \n",
       "..          ...       ...     ...  \n",
       "916           4       7.0  0.5714  \n",
       "917           4       4.0     1.0  \n",
       "918           4       4.0     1.0  \n",
       "919          61      70.0  0.8714  \n",
       "920           9      68.0  0.8676  \n",
       "\n",
       "[120882 rows x 17 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>variant</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>codon_idx</th>\n",
       "      <th>nucleotide_index</th>\n",
       "      <th>indel_length</th>\n",
       "      <th>indel_nucleotides</th>\n",
       "      <th>vcf_evidence</th>\n",
       "      <th>vcf_idx</th>\n",
       "      <th>var</th>\n",
       "      <th>is_null</th>\n",
       "      <th>is_minor</th>\n",
       "      <th>minor_variant</th>\n",
       "      <th>minor_reads</th>\n",
       "      <th>coverage</th>\n",
       "      <th>FRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1542</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>7570c&gt;t:76</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>90.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7570</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV\": [34, 76], \"COV_TOTAL\": 110, \"GT\": [1, ...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7570c&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>7570c&gt;t</td>\n",
       "      <td>76</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.6909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1543</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>7572t&gt;c:35</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>91.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7572</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV\": [77, 35], \"COV_TOTAL\": 112, \"GT_CONF\":...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7572t&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>7572t&gt;c</td>\n",
       "      <td>35</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.6875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1544</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>24699_del_ccgcgttgctcggggtaa:41</td>\n",
       "      <td>fhaA</td>\n",
       "      <td>729.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24699</td>\n",
       "      <td>-18</td>\n",
       "      <td>ccgcgttgctcggggtaa</td>\n",
       "      <td>{\"COV_TOTAL\": 103, \"COV\": [10, 41], \"GT\": [1, ...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24699_minorindel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>24699_del_ccgcgttgctcggggtaa</td>\n",
       "      <td>41</td>\n",
       "      <td>103.0</td>\n",
       "      <td>0.7767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1545</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>24721g&gt;c:7</td>\n",
       "      <td>fhaA</td>\n",
       "      <td>242.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24721</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT\": [0, 0], \"COV_TOTAL\": 85, \"FRS\": 0.9176,...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24721g&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>24721g&gt;c</td>\n",
       "      <td>7</td>\n",
       "      <td>85.0</td>\n",
       "      <td>0.9176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1546</th>\n",
       "      <td>site.ENA.subj.SAMN01908825.lab.1.iso.1</td>\n",
       "      <td>24722t&gt;c:7</td>\n",
       "      <td>fhaA</td>\n",
       "      <td>241.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>24722</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"FRS\": 0.9195, \"COV\": [73, 7], \"ALLELE_DP\": [...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24722t&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>24722t&gt;c</td>\n",
       "      <td>7</td>\n",
       "      <td>87.0</td>\n",
       "      <td>0.9195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>916</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>3948649a&gt;c:4</td>\n",
       "      <td>PE_PGRS57</td>\n",
       "      <td>952.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3948649</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV_TOTAL\": 7, \"FRS\": 0.5714, \"GT\": [1, 1], ...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3948649a&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>3948649a&gt;c</td>\n",
       "      <td>4</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.5714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>917</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>4032006c&gt;g:4</td>\n",
       "      <td>PE_PGRS58</td>\n",
       "      <td>385.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4032006</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV\": [0, 4], \"GT\": [1, 1], \"DP\": 4, \"ALLELE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4032006c&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4032006c&gt;g</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>918</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>4036926_del_gccgcccgc:4</td>\n",
       "      <td>PE_PGRS59</td>\n",
       "      <td>1117.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4036926</td>\n",
       "      <td>-9</td>\n",
       "      <td>gccgcccgc</td>\n",
       "      <td>{\"GT_CONF\": 36.62, \"COV_TOTAL\": 4, \"GT_CONF_PE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4036926_minorindel</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4036926_del_gccgcccgc</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>4060210t&gt;a:61</td>\n",
       "      <td>esxV</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4060210</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"COV\": [9, 61], \"FRS\": 0.8714, \"GT_CONF_PERCE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4060210t&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4060210t&gt;a</td>\n",
       "      <td>61</td>\n",
       "      <td>70.0</td>\n",
       "      <td>0.8714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920</th>\n",
       "      <td>site.ENA.subj.SAMN07659930.lab.1.iso.1</td>\n",
       "      <td>4263501t&gt;g:9</td>\n",
       "      <td>Rv3802c</td>\n",
       "      <td>289.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4263501</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>{\"GT_CONF\": 311.55, \"GT_CONF_PERCENTILE\": 52.8...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4263501t&gt;x</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4263501t&gt;g</td>\n",
       "      <td>9</td>\n",
       "      <td>68.0</td>\n",
       "      <td>0.8676</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18354 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    uniqueid                          variant  \\\n",
       "1542  site.ENA.subj.SAMN01908825.lab.1.iso.1                       7570c>t:76   \n",
       "1543  site.ENA.subj.SAMN01908825.lab.1.iso.1                       7572t>c:35   \n",
       "1544  site.ENA.subj.SAMN01908825.lab.1.iso.1  24699_del_ccgcgttgctcggggtaa:41   \n",
       "1545  site.ENA.subj.SAMN01908825.lab.1.iso.1                       24721g>c:7   \n",
       "1546  site.ENA.subj.SAMN01908825.lab.1.iso.1                       24722t>c:7   \n",
       "...                                      ...                              ...   \n",
       "916   site.ENA.subj.SAMN07659930.lab.1.iso.1                     3948649a>c:4   \n",
       "917   site.ENA.subj.SAMN07659930.lab.1.iso.1                     4032006c>g:4   \n",
       "918   site.ENA.subj.SAMN07659930.lab.1.iso.1          4036926_del_gccgcccgc:4   \n",
       "919   site.ENA.subj.SAMN07659930.lab.1.iso.1                    4060210t>a:61   \n",
       "920   site.ENA.subj.SAMN07659930.lab.1.iso.1                     4263501t>g:9   \n",
       "\n",
       "           gene  gene_position  codon_idx  nucleotide_index  indel_length  \\\n",
       "1542       gyrA           90.0        1.0              7570             0   \n",
       "1543       gyrA           91.0        0.0              7572             0   \n",
       "1544       fhaA          729.0        1.0             24699           -18   \n",
       "1545       fhaA          242.0        0.0             24721             0   \n",
       "1546       fhaA          241.0        2.0             24722             0   \n",
       "...         ...            ...        ...               ...           ...   \n",
       "916   PE_PGRS57          952.0        2.0           3948649             0   \n",
       "917   PE_PGRS58          385.0        0.0           4032006             0   \n",
       "918   PE_PGRS59         1117.0        2.0           4036926            -9   \n",
       "919        esxV           20.0        1.0           4060210             0   \n",
       "920     Rv3802c          289.0        0.0           4263501             0   \n",
       "\n",
       "       indel_nucleotides                                       vcf_evidence  \\\n",
       "1542                None  {\"COV\": [34, 76], \"COV_TOTAL\": 110, \"GT\": [1, ...   \n",
       "1543                None  {\"COV\": [77, 35], \"COV_TOTAL\": 112, \"GT_CONF\":...   \n",
       "1544  ccgcgttgctcggggtaa  {\"COV_TOTAL\": 103, \"COV\": [10, 41], \"GT\": [1, ...   \n",
       "1545                None  {\"GT\": [0, 0], \"COV_TOTAL\": 85, \"FRS\": 0.9176,...   \n",
       "1546                None  {\"FRS\": 0.9195, \"COV\": [73, 7], \"ALLELE_DP\": [...   \n",
       "...                  ...                                                ...   \n",
       "916                 None  {\"COV_TOTAL\": 7, \"FRS\": 0.5714, \"GT\": [1, 1], ...   \n",
       "917                 None  {\"COV\": [0, 4], \"GT\": [1, 1], \"DP\": 4, \"ALLELE...   \n",
       "918            gccgcccgc  {\"GT_CONF\": 36.62, \"COV_TOTAL\": 4, \"GT_CONF_PE...   \n",
       "919                 None  {\"COV\": [9, 61], \"FRS\": 0.8714, \"GT_CONF_PERCE...   \n",
       "920                 None  {\"GT_CONF\": 311.55, \"GT_CONF_PERCENTILE\": 52.8...   \n",
       "\n",
       "      vcf_idx                 var  is_null  is_minor  \\\n",
       "1542      1.0             7570c>x    False      True   \n",
       "1543      1.0             7572t>x    False      True   \n",
       "1544      1.0    24699_minorindel    False      True   \n",
       "1545      1.0            24721g>x    False      True   \n",
       "1546      1.0            24722t>x    False      True   \n",
       "...       ...                 ...      ...       ...   \n",
       "916       1.0          3948649a>x    False      True   \n",
       "917       1.0          4032006c>x    False      True   \n",
       "918       1.0  4036926_minorindel    False      True   \n",
       "919       1.0          4060210t>x    False      True   \n",
       "920       1.0          4263501t>x    False      True   \n",
       "\n",
       "                     minor_variant minor_reads  coverage     FRS  \n",
       "1542                       7570c>t          76     110.0  0.6909  \n",
       "1543                       7572t>c          35     112.0  0.6875  \n",
       "1544  24699_del_ccgcgttgctcggggtaa          41     103.0  0.7767  \n",
       "1545                      24721g>c           7      85.0  0.9176  \n",
       "1546                      24722t>c           7      87.0  0.9195  \n",
       "...                            ...         ...       ...     ...  \n",
       "916                     3948649a>c           4       7.0  0.5714  \n",
       "917                     4032006c>g           4       4.0     1.0  \n",
       "918          4036926_del_gccgcccgc           4       4.0     1.0  \n",
       "919                     4060210t>a          61      70.0  0.8714  \n",
       "920                     4263501t>g           9      68.0  0.8676  \n",
       "\n",
       "[18354 rows x 17 columns]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "variants[variants.is_minor]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
